pkg/ingester/ingester.go

```diff
@@ -954,41 +972,73 @@ type ctxKey int
 var pushReqCtxKey ctxKey = 1
 
 type pushRequestState struct {
-	requestSize int64
+	requestSize                  int64
+	requestDuration              time.Duration
+	acquiredCircuitBreakerPermit bool
+	pushErr                      error
+}
+
+func getPushRequestState(ctx context.Context) *pushRequestState {
+	if st, ok := ctx.Value(pushReqCtxKey).(*pushRequestState); ok {
+		return st
+	}
+	return nil
 }
 
 // StartPushRequest checks if ingester can start push request, and increments relevant counters.
 // If new push request cannot be started, errors convertible to gRPC status code are returned, and metrics are updated.
 func (i *Ingester) StartPushRequest(ctx context.Context, reqSize int64) (context.Context, error) {
-	ctx, _, err := i.startPushRequest(ctx, reqSize)
+	ctx, _, err := i.startPushRequest(ctx, reqSize, nil)
 	return ctx, err
 }
 
 func (i *Ingester) FinishPushRequest(ctx context.Context) {
-	st, ok := ctx.Value(pushReqCtxKey).(*pushRequestState)
-	if !ok {
+	st := getPushRequestState(ctx)
+	if st == nil {
 		return
 	}
-	i.finishPushRequest(st.requestSize)
+	i.inflightPushRequests.Dec()
+	if st.requestSize > 0 {
+		i.inflightPushRequestsBytes.Sub(st.requestSize)
+	}
+	if st.acquiredCircuitBreakerPermit {
+		_ = i.circuitBreaker.finishPushRequest(ctx, st.requestDuration, st.pushErr)
+	}
 }
 
 // This method can be called in two ways: 1. Ingester.PushWithCleanup, or 2. Ingester.StartPushRequest via gRPC server's method limiter.
 //
 // In the first case, returned errors can be inspected/logged by middleware. Ingester.PushWithCleanup will wrap the error in util_log.DoNotLogError wrapper.
 // In the second case, returned errors will not be logged, because request will not reach any middleware.
 //
-// The shouldFinish flag tells if the caller must call finish on this request. If not, there is already someone in the call stack who will do that.
-func (i *Ingester) startPushRequest(ctx context.Context, reqSize int64) (_ context.Context, shouldFinish bool, err error) {
+// Parameter finishPushRequestFn is used for testing purposes, and represents the function that finishes the push request that startPushRequest initialize,
```

Comment:
We should test with real function whenever possible, and there is nothing in real function that's difficult to test. Let's please remove this function parameter.

Reply:
I have removed the new parameter and modified the test to test the real function.